### PR TITLE
Add `dataset.loader.as_stacked_batches`.

### PR DESCRIPTION
**Pull request by jaheba**
*Description of changes:*

Add a new function to do collecting and stacking data batches. The function can be used to simplify our data loaders in the future.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Please tag this pr with at least one of these labels to make our release process faster:** BREAKING, new feature, bug fix, other change, dev setup

### CODE REVIEW

File: src/gluonts/dataset/loader.py [hunk 1]
```diff
@@ -37,6 +50,63 @@ def __call__(self, data, is_train):
...
+    def __call__(self, data, is_train):
+        for batch in data:
+            yield rows_to_columns(batch, np.array)
+
+
+def as_stacked_batches(
+    dataset: Dataset,
+    *,
+    batch_size: int,
+    output_type: Callable,
```
Comment (jaheba) [inline src/gluonts/dataset/loader.py]: Should we make this `Optional`?

File: src/gluonts/dataset/loader.py [hunk 2]
```diff
@@ -37,6 +50,63 @@ def __call__(self, data, is_train):
...
+    if shuffle_buffer_length:
+        dataset = PseudoShuffled(dataset, shuffle_buffer_length)
+
+    transform: Transformation = Identity()
+
+    if field_names is not None:
+        transform += SelectFields(field_names)
+
+    transform += Batch(batch_size=batch_size)
+    transform += Stack()
```
Comment (jaheba) [inline src/gluonts/dataset/loader.py:95]: This could also be an optional parameter, and only stack if required.

On the other hand, I don't know where we would want that.

File: src/gluonts/dataset/loader.py [hunk 3]
```diff
@@ -37,6 +50,63 @@ def __call__(self, data, is_train):
         yield from batcher(data, self.batch_size)
 
 
+class Stack(Transformation, BaseModel):
+    def __call__(self, data, is_train):
+        for batch in data:
+            yield rows_to_columns(batch, np.array)
```
Comment (jaheba) [inline src/gluonts/dataset/loader.py:56]: There is an argument that we don't need to wrap this in an `np.array`.

However, `torch.tensor` and `mx.ndarray` like to work with `np.array` instances rather than lists.

File: src/gluonts/dataset/loader.py (hunk 2 above)
Comment (lostella) [reply src/gluonts/dataset/loader.py:95]: yeah no need for that

File: src/gluonts/dataset/loader.py (hunk 1 above)
Comment (lostella) [reply src/gluonts/dataset/loader.py]: Why not, or just default to identity?

File: src/gluonts/dataset/loader.py (hunk 3 above)
Comment (lostella) [reply src/gluonts/dataset/loader.py:56]: You could wrap as numpy array in a separate step using `ValMap`? But then again, it's also fine like this

Comment (jaheba) [reply src/gluonts/dataset/loader.py:56]: I guess we can leave it as it is.

File: src/gluonts/dataset/loader.py [hunk 4]
```diff
@@ -37,6 +50,61 @@ def __call__(self, data, is_train):
...
+
+    if output_type is not None:
+        transform += Valmap(output_type)
+
+    # Note: is_train needs to be provided but does not have an effect
+    transformed_dataset = transform.apply(dataset, is_train=True)
+    return IterableSlice(
+        transformed_dataset,
+        num_batches_per_epoch,
+    )
```
Comment (lostella) [inline src/gluonts/dataset/loader.py]: If this creates complications (we may or may not need to `iter` the `transformed_dataset`) we could even remove this wrapping and rely on the trainer to limit how much data is taken at once, see the `limit_train_batches` option at https://pytorch-lightning.readthedocs.io/en/stable/api/pytorch_lightning.trainer.trainer.Trainer.html#pytorch_lightning.trainer.trainer.Trainer

This way we can directly return `transformed_dataset` and downstream one `iter`s it or not depending on the context

Comment (jaheba) [reply src/gluonts/dataset/loader.py]: I don't think this causes issues. One is free to not set `num_batches_per_epoch`.

Comment (lostella) [reply src/gluonts/dataset/loader.py]: it prevents you from being able to `iter(transformed_dataset)` downstream

Comment (lostella) [reply src/gluonts/dataset/loader.py]: nvm

**Review by lostella — APPROVED** [review]
